Source/WebCore/dom/Document.cpp

```diff
@@ -2097,6 +2099,54 @@ std::optional<BoundaryPoint> Document::caretPositionFromPoint(const LayoutPoint&
     return { { node.releaseNonNull(), offset } };
 }
 
+RefPtr<CaretPosition> Document::caretPositionFromPoint(double x, double y, CaretPositionFromPointOptions options)
+{
+    if (!hasLivingRenderTree())
+        return nullptr;
+
+    LayoutPoint clientPoint = { x, y };
+
+    LayoutPoint localPoint;
+    RefPtr node = nodeFromPoint(clientPoint, &localPoint, HitTestSource::Script);
+    if (!node)
+        return nullptr;
+
+    updateLayoutIgnorePendingStylesheets();
+
+    CheckedPtr renderer = node->renderer();
+    if (!renderer)
+        return nullptr;
+
+    if (renderer->isSkippedContentRoot())
+        return CaretPosition::create(WTFMove(node), 0);
+
+    auto position = renderer->positionForPoint(localPoint, HitTestSource::Script);
+    if (position.isNull())
+        return nullptr;
+
+    RefPtr anchorNode = position.anchorNode();
+    if (anchorNode->shadowHost()) {
+        if (auto* textFormControl = dynamicDowncast<HTMLTextFormControlElement>(*anchorNode->shadowHost())) {
```

Comment:
```suggestion
    if (auto* textFormControl = dynamicDowncast<HTMLTextFormControlElement>(anchorNode->shadowHost())) {
```
Leaving out the `*` allows `dynamicDowncast` to do the null check for us.

Reply:
Clever, fixed.